secure_storage: fix Kconfig dependencies

JordanYates opened the pull request:
`SECURE_STORAGE_ITS_STORE_IMPLEMENTATION_SETTINGS` does not use either the `FLASH_MAP` or `NVS` APIs, only `SETTINGS`. Similarly, `SECURE_STORAGE_ITS_STORE_MODULE` does not consist of any code itself and therefore should not select or imply any options.

Mixing `select` and `depends on` is a common source of Kconfig dependency loops and should be avoided. Both `ZMS` and `SETTINGS` are more commonly used with `depends on` rather than `select`.

The usage here also contradicts the [Zephyr best practices](https://docs.zephyrproject.org/latest/build/kconfig/tips.html#select-recommendations) guide for `select`:
 * Avoid selecting symbols with prompts or dependencies. Prefer depends on.

Fixes #86328

JordanYates committed:
secure_storage: remove incorrect `imply` symbols

`SECURE_STORAGE_ITS_STORE_IMPLEMENTATION_SETTINGS` does not use either
the `FLASH_MAP` or `NVS` APIs, only `SETTINGS`.

Similarly, `SECURE_STORAGE_ITS_STORE_MODULE` does not consist of any
code itself and therefore should not select or imply any options.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
secure_storage: swap `select` to `depends on`

Mixing `select` and `depends on` is a common source of Kconfig
dependency loops and should be avoided. Both `ZMS` and `SETTINGS` are
more commonly used with `depends on` rather than `select`.

The usage here also contradicts the Zephyr best practices guide for
`select`:
 * Avoid selecting symbols with prompts or dependencies.
   Prefer depends on.

Signed-off-by: Jordan Yates <jordan@embeint.com>

rruuaanng reviewed (dismissed):
Looks good.

JordanYates commented:
Removed additional `imply` on `SECURE_STORAGE_ITS_STORE_MODULE`

tomi-font reviewed (changes requested):
Thanks for the fixes.

You will need to adjust the configurations under `samples/psa/` and `tests/subsys/secure_storage/` to enable those Kconfig options where needed.
Because now with your PR all the tests get skipped because of the filters as some conditions are missing:
```
zephyr$ twister -p native_sim -T ./samples/psa
[...]
INFO    - 1/6 native_sim/native         samples/psa/its/sample.psa.its.secure_storage.entropy_driver FILTERED (runtime filter)
INFO    - 2/6 native_sim/native         samples/psa/persistent_key/sample.psa.persistent_key.secure_storage.entropy_not_secure FILTERED (runtime filter)
INFO    - 3/6 native_sim/native         samples/psa/persistent_key/sample.psa.persistent_key.secure_storage.entropy_driver FILTERED (runtime filter)
INFO    - 4/6 native_sim/native         samples/psa/its/sample.psa.its.tfm                 FILTERED (runtime filter)
INFO    - 5/6 native_sim/native         samples/psa/persistent_key/sample.psa.persistent_key.tfm FILTERED (runtime filter)
INFO    - 6/6 native_sim/native         samples/psa/its/sample.psa.its.secure_storage.entropy_not_secure FILTERED (runtime filter)
```

JordanYates committed:
tests: secure_storage: add dependencies

Now that `SECURE_STORAGE` does not `select` dependencies, they need to
be enabled explicitly by the tests.

Signed-off-by: Jordan Yates <jordan@embeint.com>

tomi-font commented on subsys/secure_storage/Kconfig.its_store:
```diff
@@ -31,8 +31,6 @@ config SECURE_STORAGE_ITS_STORE_IMPLEMENTATION_SETTINGS
 	     && $(dt_node_has_compat,$(dt_node_parent,$(DT_SETTINGS_PARTITIION)),fixed-partitions))\
 	  || ($(dt_path_enabled,$(DT_STORAGE_PARTITION)) \
 	      && $(dt_node_has_compat,$(dt_node_parent,$(DT_STORAGE_PARTITION)),fixed-partitions)))
-	imply FLASH_MAP
-	imply NVS
```
As said that was to enable `SETTINGS_NVS` by default but allow that to be changed to the other options you mention.

JordanYates replied:
Adding `imply` to a symbol should only be done if absolutely necessary, since the only way to disable that is by `CONFIG_X=n` in `prj.conf`.

There is no way for a Zephyr module or downstream to specify they want to use a different default and `unimply` the choice you have made here.

tomi-font commented on tests/subsys/secure_storage/psa/its/overlay-store_default.conf:
This file could actually be renamed to `overlay-store_settings.conf`.

tomi-font reviewed (commented):
This change would also deserve a migration guide entry.

JordanYates committed:
fs: nvs: move `FLASH_PAGE_LAYOUT` to `depends on`

`FLASH_PAGE_LAYOUT` has a hardware dependency on `FLASH_HAS_PAGE_LAYOUT`
which is not present for all boards. Forcing this symbol to `y` when
the hardware doesn't support it results in build errors at the Kconfig
stage.

`FLASH_PAGE_LAYOUT` is enabled by default when `FLASH_HAS_PAGE_LAYOUT`
is true, so this change will not require any user changes.

Signed-off-by: Jordan Yates <jordan@embeint.com>

tomi-font reviewed (changes requested):
Just a few nits left.

tomi-font commented on doc/releases/migration-guide-4.1.rst:
```diff
@@ -677,6 +677,14 @@ LoRa
+* Backends no longer automatically enable their dependencies through ``select`` or ``imply``. Users
+  must ensure that the depencies are enabled in their applications. For example,
+  :kconfig:option:`SECURE_STORAGE_ITS_STORE_IMPLEMENTATION_SETTINGS` requires
+  :kconfig:option:`CONFIG_SETTINGS` to be enabled elsewhere.
```
My point here was to explicitly state that `CONFIG_SECURE_STORAGE_ITS_STORE_IMPLEMENTATION_SETTINGS` would automatically enable the *NVS* settings backend, so that the users are aware of what used to be the default if they hadn't changed it.

JordanYates replied:
Can you provide the wording you want? Anything I can think of is sort of redundant due to the "no longer automatically enable their dependencies" in the first sentence. Seems kind of self explanatory to me.

tomi-font commented on doc/releases/migration-guide-4.1.rst:
```diff
@@ -677,6 +677,15 @@ LoRa
+
+* Store backends no longer automatically enable their dependencies through ``select`` or ``imply``.
+  Users must ensure that the depencies are enabled in their applications. For example,
+  :kconfig:option:`SECURE_STORAGE_ITS_STORE_IMPLEMENTATION_SETTINGS` requires
```
You're missing the `CONFIG_` prefix on two Kconfig options:
```suggestion
  :kconfig:option:`CONFIG_SECURE_STORAGE_ITS_STORE_IMPLEMENTATION_SETTINGS` requires
```

tomi-font commented on doc/releases/migration-guide-4.1.rst:
```diff
@@ -677,6 +677,15 @@ LoRa
+==============
+
+* Store backends no longer automatically enable their dependencies through ``select`` or ``imply``.
+  Users must ensure that the depencies are enabled in their applications. For example,
+  :kconfig:option:`SECURE_STORAGE_ITS_STORE_IMPLEMENTATION_SETTINGS` requires
+  :kconfig:option:`CONFIG_SETTINGS` and :kconfig:option:`SECURE_STORAGE_ITS_STORE_IMPLEMENTATION_ZMS`
+  requires :kconfig:option:`CONFIG_ZMS` (:github:`86181`).
```
Something like
```suggestion
  Users must now ensure that the dependencies are enabled in their applications.
  :kconfig:option:`CONFIG_SECURE_STORAGE_ITS_STORE_IMPLEMENTATION_SETTINGS` previously enabled NVS and settings, which means the NVS settings backend would get used by default if ZMS wasn't enabled. (:github:`86181`)
```
